Allow spare nodes to join and leave an OpenMPF cluster while the WFM is running

jrobble opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/openmpf/openmpf/517)
<!-- Reviewable:end -->

jrobble commented:
Reviewed 15 of 39 files at r1, 9 of 24 files at r2, 38 of 38 files at r3.
Review status: 53 of 60 files reviewed at latest revision, all discussions resolved.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 123 at r3](https://reviewable.io/reviews/openmpf/openmpf/517#-LDDz1-e2MnjKN72M3EN:-LDDz1-e2MnjKN72M3EO:bz40ch3) ([raw file](https://github.com/openmpf/openmpf/blob/a630508e0b6262f56bfd359ab0ad20e1e1cb8328/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L123)):*
> ```Python
> 
>         # ALL_MPF_NODES is deprecated. For backwards compatibility with deployments initially installed with < R2.1.0,
>         # use ALL_MPF_NODES if CORE_MPF_NODES is not defined.
> ```

TODO: Check if we really need to be backwards-compatible with ALL_MPF_NODES.

---

*[trunk/node-manager/src/main/java/org/mitre/mpf/nms/BaseServiceLauncher.java, line 507 at r3](https://reviewable.io/reviews/openmpf/openmpf/517#-LDDrTFS9AwaqtV7BHGR:-LDDrTFT3gZWrgPWT90t:be94gun) ([raw file](https://github.com/openmpf/openmpf/blob/a630508e0b6262f56bfd359ab0ad20e1e1cb8328/trunk/node-manager/src/main/java/org/mitre/mpf/nms/BaseServiceLauncher.java#L507)):*
> ```Java
>         isRunning = false;
>         mRunToCompletion = true;
>         LOG.info("RUN method for service {} ending", this.mServiceDesc.getName());
> ```

TODO: Convert info() statements back to debug().

---

*[trunk/node-manager/src/main/java/org/mitre/mpf/nms/ChannelNode.java, line 159 at r3](https://reviewable.io/reviews/openmpf/openmpf/517#-LDDrb8H5oRNeRUmWcGW:-LDDrb8H5oRNeRUmWcGX:be94gun) ([raw file](https://github.com/openmpf/openmpf/blob/a630508e0b6262f56bfd359ab0ad20e1e1cb8328/trunk/node-manager/src/main/java/org/mitre/mpf/nms/ChannelNode.java#L159)):*
> ```Java
>      */
>     public void shutdown() {
>         log.info("Shutdown requested.");
> ```

TODO: Convert info() statements back to debug().

---

*[trunk/node-manager/src/main/java/org/mitre/mpf/nms/ChannelReceiver.java, line 89 at r3](https://reviewable.io/reviews/openmpf/openmpf/517#-LDDreLxAxY4FekzJ24c:-LDDreLxAxY4FekzJ24d:be94gun) ([raw file](https://github.com/openmpf/openmpf/blob/a630508e0b6262f56bfd359ab0ad20e1e1cb8328/trunk/node-manager/src/main/java/org/mitre/mpf/nms/ChannelReceiver.java#L89)):*
> ```Java
>     @Override
>     public void getState(OutputStream output) throws Exception {
>         LOG.info("Responding to cluster state request");
> ```

TODO: Convert info() statements back to debug().

---

*[trunk/node-manager/src/main/java/org/mitre/mpf/nms/ChildNodeStateManager.java, line 77 at r3](https://reviewable.io/reviews/openmpf/openmpf/517#-LDDrj4YEREL4y8brWQP:-LDDrj4Z9gawJ3eErqxo:be94gun) ([raw file](https://github.com/openmpf/openmpf/blob/a630508e0b6262f56bfd359ab0ad20e1e1cb8328/trunk/node-manager/src/main/java/org/mitre/mpf/nms/ChildNodeStateManager.java#L77)):*
> ```Java
>         }
>         else if (obj instanceof ServiceStatusUpdate) {
>             LOG.info("Received a ServiceStatusUpdate from {}", sender);
> ```

TODO: Convert info() statements back to debug().

---

*[trunk/node-manager/src/main/java/org/mitre/mpf/nms/GenericServiceLauncher.java, line 64 at r3](https://reviewable.io/reviews/openmpf/openmpf/517#-LDDrll48dSRtC3UD928:-LDDrll48dSRtC3UD929:be94gun) ([raw file](https://github.com/openmpf/openmpf/blob/a630508e0b6262f56bfd359ab0ad20e1e1cb8328/trunk/node-manager/src/main/java/org/mitre/mpf/nms/GenericServiceLauncher.java#L64)):*
> ```Java
>     @Override
>     public void sendShutdownToApp() {
>         LOG.info("Sending down 'q' to {}", this.getServiceName());
> ```

TODO: Convert info() statements back to debug().

---

*[trunk/node-manager/src/main/java/org/mitre/mpf/nms/MasterNode.java, line 138 at r3](https://reviewable.io/reviews/openmpf/openmpf/517#-LDDrny_7pISOLnSDlvv:-LDDrny_7pISOLnSDlvw:be94gun) ([raw file](https://github.com/openmpf/openmpf/blob/a630508e0b6262f56bfd359ab0ad20e1e1cb8328/trunk/node-manager/src/main/java/org/mitre/mpf/nms/MasterNode.java#L138)):*
> ```Java
>             synchronized (nodeStateManager.getServiceTable()) {
>                 for (ServiceDescriptor service : nodeStateManager.getServiceTable().values()) {
>                     log.info("Service with name '{}' is part of the service table, but no longer part of the node manager config.",
> ```

TODO: Convert info() statements back to debug().

---

*[trunk/node-manager/src/main/java/org/mitre/mpf/nms/MasterNodeStateManager.java, line 136 at r3](https://reviewable.io/reviews/openmpf/openmpf/517#-LDDrqtV5k-53P_BVahq:-LDDrqtV5k-53P_BVahr:be94gun) ([raw file](https://github.com/openmpf/openmpf/blob/a630508e0b6262f56bfd359ab0ad20e1e1cb8328/trunk/node-manager/src/main/java/org/mitre/mpf/nms/MasterNodeStateManager.java#L136)):*
> ```Java
>                         break;
>                     default:
>                         LOG.info("Unprocessed Manager StatusUpdate state: {} ", state);
> ```

TODO: Convert info() statements back to debug().

---

*[trunk/node-manager/src/main/java/org/mitre/mpf/nms/OutputShredder.java, line 182 at r3](https://reviewable.io/reviews/openmpf/openmpf/517#-LDDs7xw-MjcSNMVCGGc:-LDDs7xx32KU-I59xocq:be94gun) ([raw file](https://github.com/openmpf/openmpf/blob/a630508e0b6262f56bfd359ab0ad20e1e1cb8328/trunk/node-manager/src/main/java/org/mitre/mpf/nms/OutputShredder.java#L182)):*
> ```Java
>                 }
>             } catch (IOException ex) {
>                 LOG.info("IOException reading from {}: ", outputName, ex);
> ```

TODO: Convert info() statements back to debug().

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/util/PropertiesUtil.java, line 140 at r3](https://reviewable.io/reviews/openmpf/openmpf/517#-LDDyoIQ3yGXVzzUDA0E:-LDDyoIR06XGg41FpvAM:bz40ch3) ([raw file](https://github.com/openmpf/openmpf/blob/a630508e0b6262f56bfd359ab0ad20e1e1cb8328/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/util/PropertiesUtil.java#L140)):*
> ```Java
> 
>     private void parseCoreMpfNodes() {
>         // NOTE: Switched from using ALL_MPF_NODES to CORE_MPF_NODES in R2.1.0. Use CORE_MPF_NODES if possible.
> ```

TODO: Check if we really need to be backwards-compatible with ALL_MPF_NODES.

---


*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517)*
<!-- Sent from Reviewable.io -->

jrobble commented:
Reviewed 7 of 9 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517)*
<!-- Sent from Reviewable.io -->

jrobble commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[trunk/ansible/install/ansible/roles/master/templates/mpf-profile.sh.j2, line 38 at r4](https://reviewable.io/reviews/openmpf/openmpf/517#-LDDzfXvATmkDfdqPlSb:-LDDzfXvATmkDfdqPlSc:b7p20av) ([raw file](https://github.com/openmpf/openmpf/blob/1a90e711c6cccc706a487c5a218611c2fa466070/trunk/ansible/install/ansible/roles/master/templates/mpf-profile.sh.j2#L38)):*
> ```
> export JAVA_HOME=/usr/java/latest
> 
> export JGROUPS_TCP_ADDRESS=$THIS_MPF_NODE
> ```

TODO: Perform a full deployment.

---


*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517)*
<!-- Sent from Reviewable.io -->

jrobble commented:
To test:

1. Deploy OpenMPF to multi-node cluster. Decide which non-master node to make spare.

2. When WFM is running, remove spare node using Nodes UI.

3. `mpf stop`

4. `sudo vi /etc/profile.d/mpf.sh`
    Remove spare node from CORE_MPF_NODES list.

5. `sudo vi /etc/ansible/hosts`
   Remove spare node from [mpf-child] list.

6. Stop spare node VM.

7. `mpf start`
   Verify that core nodes appear in Nodes UI and services are running.
   Verfiy that "Add Node" button on Nodes UI shows a dialog that states that all available nodes have been configured.

8. `mpf list-nodes`
   Verify that only remaining core nodes are listed.

8. Start spare node VM. Wait 1 minute.

9. `mpf list-nodes`
   Verify that remaining core nodes and spare node is listed.

10. Use "Add Node" button on Nodes UI to add spare node. Add services.
    Verify that spare node is labeled "Spare" and not "Core".
    Verify that services are running on the spare node and can be used to execute a job.
    Verify on the Component Registration page that "Note that all actions performed on this page will only affect core nodes:" only lists the remaining core nodes.

11. Remove a component. Verify that it succeeds.

12. Register a component. Verify that it succeeds.
    Verify that you can run a job with this component.
    If this is a new component, verify that if you configure the spare node to run this component service, the service will fail.

---

Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517#-:-LEY4Wz32Qkpyv3Mv54B:b-hhbgn8)*
<!-- Sent from Reviewable.io -->

jrobble commented:
Reviewed 7 of 8 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517)*
<!-- Sent from Reviewable.io -->

jrobble commented:
*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 99 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnq6cE2f56oONbpi1s:-LEpI3gfCcWL96ZlywoJ:bupqjfh) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L99)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

You should use the built in json library to parse this response. So this would end up as `return json.load(response)`
</blockquote></details>

I originally did that. Here's what it shows:

```
Core nodes: [u'10.0.2.101', u'10.0.2.100', u'10.0.2.102']
No spare nodes
```

Here's what I want:

```
Core nodes: ['10.0.2.101,10.0.2.100,10.0.2.102']
No spare nodes
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517)*
<!-- Sent from Reviewable.io -->

jrobble commented:
*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 99 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnq6cE2f56oONbpi1s:-LEpPNrl1cuCSOuynLYS:bjwnzcm) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L99)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

I originally did that. Here's what it shows:

```
Core nodes: [u'10.0.2.101', u'10.0.2.100', u'10.0.2.102']
No spare nodes
```

Here's what I want:

```
Core nodes: ['10.0.2.101,10.0.2.100,10.0.2.102']
No spare nodes
```
</blockquote></details>

I realized that what I provided as my desired output is actually not what I want. I want this:

```
Core nodes: ['10.0.2.101', '10.0.2.100', '10.0.2.102']
No spare nodes
```
This:

```
return response.read()[2:-2].translate(None, '"').split()
```

Should be:

```
return response.read()[2:-2].translate(None, '"').split(',')
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517)*
<!-- Sent from Reviewable.io -->

jrobble commented:
Reviewed 2 of 2 files at r9.
Review status: all files reviewed, 24 unresolved discussions (waiting on @brosenberg42 and @jrobble)

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 114 at r7](https://reviewable.io/reviews/openmpf/openmpf/517#-LEk32CWDViR2rNbNIzg:-LEpQXe7BvBSmKk1fDKs:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/2b4943534faac915400a80e8fb8f03d82b266749/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L114)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

You should get the value of `$CORE_MPF_NODES` using `os.environ`. This method won't handle variable interpolation properly and trunk/mpf-install/src/main/scripts/mpf-profile.sh sets `export CORE_MPF_NODES=$THIS_MPF_NODE`
</blockquote></details>

Done.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 29 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnrSVH9rtYZ4sUXdeP:-LEpQfIa6PNbYuP_QZF2:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L29)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

This is unused
</blockquote></details>

Done.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 43 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnjD6lBJJQ5eg2XaVU:-LEpQmle-rWD7heEj834:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L43)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

This should be declared as:
```
@argh.arg('-w', '--workflow-manager-url', help='Url to Workflow Manager')
def list_nodes(workflow_manager_url='http://localhost:8080/workflow-manager'):
```
argh will check if the parameter has a default value and add that to the help message. 
</blockquote></details>

Done.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 56 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnlHNl-JFD6oSZAeVH:-LEpQnr0A25ia2PX_PtD:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L56)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

When this error occurs, the process returns a success exit code. This should raise `mpf_util.MpfError`. command_line.py handles `mpf_util.MpfError` by printing the message and exiting with a non-success exit code
</blockquote></details>

Done.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 61 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnmRE30iZJtIn6gjIU:-LEpQocoEWEbkX0NTyKd:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L61)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

There should not be `[` and `]` here.
It should just be: `username, password = get_username_and_password()`
</blockquote></details>

Done.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 65 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnmtgN0xuUV3lBURnz:-LEpQpqF1aphF3Z5Bzyr:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L65)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

This file mixes `'` and `"`.  It should only use `'` since that is what is used in the rest of these scripts.
</blockquote></details>

Done.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 73 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnnFsJ0eDjoT6f_tjD:-LEpQqv30bmcgCTd2fdr:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L73)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

`for_admin` is always `False`.
</blockquote></details>

Done.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 82 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnnNCF4Mm9CzmJzPBq:-LEpQrcy0X_Sou16Qo7U:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L82)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

Normally you don't put `[` and `]` around multiple return values. This should just be:  `return username, password`
</blockquote></details>

Done.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 85 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnnhGW1L22doG7-QaL:-LEpQsYhAcIJoiWhfk00:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L85)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

The last argument, `type` shadows a common builtin so I would change it to `node_type`.
</blockquote></details>

Done.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 86 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnoY8b40o2TrP76kJd:-LEpQw-e-aUDvfhsVnnN:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L86)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

`str.join` is used to convert an iterable to str. This is sort of weird because you combine join and +. 
This should be:
` endpoint_url = wfm_manager_url.rstrip('/') + '/rest/nodes/all?type=' + node_type`

Also you should use `wfm_manager_url.rstrip('/')` instead of `string.rstrip`. `string.rstrip` is listed as a deprecated function in https://docs.python.org/2/library/string.html#deprecated-string-functions
</blockquote></details>

Done.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 89 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEno7Kc5CzQA1DneG-I:-LEpQxiG--YauCflwm4G:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L89)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

This is not necessary. It is really a hack to get it to send a `HEAD` request instead of `GET` or `POST`
</blockquote></details>

Done.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 109 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnqoOn9RWIg3cbt_ti:-LEpQyiG8t4USvom19Z0:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L109)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

You should only catch `URLError` here
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517)*
<!-- Sent from Reviewable.io -->

brosenberg42 commented:
*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 99 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnq6cE2f56oONbpi1s:-LEpQIfC9oVbI-Aie1IT:b-5t8fd) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L99)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

I realized that what I provided as my desired output is actually not what I want. I want this:

```
Core nodes: ['10.0.2.101', '10.0.2.100', '10.0.2.102']
No spare nodes
```
This:

```
return response.read()[2:-2].translate(None, '"').split()
```

Should be:

```
return response.read()[2:-2].translate(None, '"').split(',')
```
</blockquote></details>

I don't want your code to break if the format of the JSON from the server slightly changes.
To get rid of the u, use:
```
print '[' +  ', '.join(nodes_list) + ']'
```
Sure, converting to your desired format is a little bit more complicated, but at least it handles all valid JSON.

---


*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517)*
<!-- Sent from Reviewable.io -->

brosenberg42 commented:
Review status: all files reviewed, 17 unresolved discussions (waiting on @jrobble and @brosenberg42)

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 42 at r9](https://reviewable.io/reviews/openmpf/openmpf/517#-LEpSaGy5oC9i33RWcoU:-LEpSaGz3IMDD9B_RDLl:b-ue8s9q) ([raw file](https://github.com/openmpf/openmpf/blob/0f9d1d41687254fca49e46622e23e9eea4aede83/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L42)):*
> ```Python
>         print mpf_util.MsgUtil.yellow('Cannot determine live JGroups membership.')
> 
>         core_mpf_nodes_str = os.environ.get(CORE_MPF_NODES_ENV_VAR).strip(', ')
> ```

`os.environ.get(CORE_MPF_NODES_ENV_VAR)` will return null if the environment variable is not set.
I think you want `os.environ.get(CORE_MPF_NODES_ENV_VAR, '')`

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 54 at r9](https://reviewable.io/reviews/openmpf/openmpf/517#-LEpTVP9CsG-4T22F7Wa:-LEpTVP9CsG-4T22F7Wb:bbs9647) ([raw file](https://github.com/openmpf/openmpf/blob/0f9d1d41687254fca49e46622e23e9eea4aede83/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L54)):*
> ```Python
>     username, password = get_username_and_password()
>     core_nodes_list = get_all_wfm_nodes(workflow_manager_url, username, password, "core")
>     spare_nodes_list = get_all_wfm_nodes(workflow_manager_url, username, password, "spare")
> ```

Change `"` to `'` in the two lines aboe.

---


*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517)*
<!-- Sent from Reviewable.io -->

jrobble commented:
Reviewed 4 of 5 files at r8, 10 of 10 files at r10.
Review status: all files reviewed, 17 unresolved discussions (waiting on @brosenberg42)

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 99 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnq6cE2f56oONbpi1s:-LEpWe3s4mNjpp1HUBT2:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L99)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

I don't want your code to break if the format of the JSON from the server slightly changes.
To get rid of the u, use:
```
print '[' +  ', '.join(nodes_list) + ']'
```
Sure, converting to your desired format is a little bit more complicated, but at least it handles all valid JSON.
</blockquote></details>

Done.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 42 at r9](https://reviewable.io/reviews/openmpf/openmpf/517#-LEpSaGy5oC9i33RWcoU:-LEpWj4E4_mw76WPm3VE:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/0f9d1d41687254fca49e46622e23e9eea4aede83/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L42)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

`os.environ.get(CORE_MPF_NODES_ENV_VAR)` will return null if the environment variable is not set.
I think you want `os.environ.get(CORE_MPF_NODES_ENV_VAR, '')`
</blockquote></details>

Done.

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 54 at r9](https://reviewable.io/reviews/openmpf/openmpf/517#-LEpTVP9CsG-4T22F7Wa:-LEpWrQ5-y9k5FurDZSX:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/0f9d1d41687254fca49e46622e23e9eea4aede83/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L54)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

Change `"` to `'` in the two lines aboe.
</blockquote></details>

Done.

---

*[trunk/node-manager/src/main/java/org/mitre/mpf/nms/AddressParser.java, line 59 at r7](https://reviewable.io/reviews/openmpf/openmpf/517#-LEkcMyj07KuQaJUYP1K:-LEpXlTY8Q6H0raGsXIN:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/2b4943534faac915400a80e8fb8f03d82b266749/trunk/node-manager/src/main/java/org/mitre/mpf/nms/AddressParser.java#L59)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

This is a little bit confusing, I would either add a comment stating that we want to see the stack trace or just log the exception without throwing it.
This will still print the stack trace:
```
IllegalStateException ex = new IllegalStateException("Address " + address + " is not in the form <node-type>" + FQN_SEP + "<host>.....")
LOG.warn(ex.getMessage(), e)
```
</blockquote></details>

Done.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/mvc/controller/AdminLogsController.java, line 69 at r7](https://reviewable.io/reviews/openmpf/openmpf/517#-LEk5fKJBPKj9KSadhgW:-LEpXu7o34jnh5zanwey:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/2b4943534faac915400a80e8fb8f03d82b266749/trunk/workflow-manager/src/main/java/org/mitre/mpf/mvc/controller/AdminLogsController.java#L69)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

`thisHost` is unused
</blockquote></details>

Done.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/mvc/controller/AdminPropertySettingsController.java, line 103 at r7](https://reviewable.io/reviews/openmpf/openmpf/517#-LEkgGt87eorcJI0rv5f:-LEp_3xR8V412i3MoTKM:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/2b4943534faac915400a80e8fb8f03d82b266749/trunk/workflow-manager/src/main/java/org/mitre/mpf/mvc/controller/AdminPropertySettingsController.java#L103)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

This should probably return a non-success status code. Returning null will still result in 200 status. You can just throw an exception which will be handled by `ControllerUncaughtExceptionHandler` or you can return a `ResponseEntity<...>` if you want a specific error code.
</blockquote></details>

Done.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/mvc/controller/NodeController.java, line 154 at r7](https://reviewable.io/reviews/openmpf/openmpf/517#-LEkfvkeEzzBdtc8V-4s:-LEpcInQ2MEatABR6JaQ:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/2b4943534faac915400a80e8fb8f03d82b266749/trunk/workflow-manager/src/main/java/org/mitre/mpf/mvc/controller/NodeController.java#L154)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

You call `nodes.addAll(allAvailableNodes);` in both branches here.

This part can be simplified to:
```
Set<String> nodes = new HashSet<>(nodeManagerService.getAvailableNodes());
if (type.equals("spare")) {
    nodes.removeAll(coreNodes);
    return nodes;
}
if (type.equals("all")) {
    nodes.addAll(coreNodes);
    return nodes;
}
```
</blockquote></details>

Done.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/mvc/controller/NodeController.java, line 171 at r7](https://reviewable.io/reviews/openmpf/openmpf/517#-LEkhDJEDvX_z6PTex0Y:-LEpaJ6sDio-Tx54ybfI:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/2b4943534faac915400a80e8fb8f03d82b266749/trunk/workflow-manager/src/main/java/org/mitre/mpf/mvc/controller/NodeController.java#L171)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

This should probably return a non-success status code. Returning null will still result in 200 status. You can just throw an exception which will be handled by `ControllerUncaughtExceptionHandler` or you can return a `ResponseEntity<...>` if you want a specific error code.
</blockquote></details>

Done.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/enums/EnvVar.java, line 29 at r7](https://reviewable.io/reviews/openmpf/openmpf/517#-LEk6bCF0Dys3T0QVEi6:-LEpaUeq5gbwcpJWM8hC:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/2b4943534faac915400a80e8fb8f03d82b266749/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/enums/EnvVar.java#L29)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

Classes with only static fields should have a private constructor.
</blockquote></details>

Done.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/NodeManagerServiceImpl.java, line 173 at r7](https://reviewable.io/reviews/openmpf/openmpf/517#-LEknpsKAZ0KN-RmgeGI:-LEpbibdAL4pb3IgTYpp:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/2b4943534faac915400a80e8fb8f03d82b266749/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/NodeManagerServiceImpl.java#L173)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

The `convertToModel` method was added so that it was possible to create the `NodeManagerModel` in one line. This should be changed so that `NodeManagerModel.setOnline` is called in `convertToModel`. You will also need to add a parameter to `convertToModel` so you can pass in the `availableNodes` set. In the end this should look like:
```
Set<String> availableNodes = getAvailableNodes();
return managers.managers()
        .stream()
        .map(m -> convertToModel(m, availableNodes))
        .collect(toCollection(ArrayList::new));
```
</blockquote></details>

Done.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/util/PropertiesUtil.java, line 82 at r7](https://reviewable.io/reviews/openmpf/openmpf/517#-LEkhL2VDjZJ3EJ-SLlX:-LEpbmVV9CN2Dta9zThJ:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/2b4943534faac915400a80e8fb8f03d82b266749/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/util/PropertiesUtil.java#L82)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

This collection gets directly returned. So I think an `ImmutableSet<String>` would be better here.  If not, I would at least wrap the HashSet in `Collections.unmodifiableSet`.
I trying to avoid bugs where someone accidentally modifies this collection for example:
```
Set<String> coreNodes = nodeManagerService.getCoreNodes();
if (type.equals("all")) {
    coreNodes.addAll(nodeManagerService.getAvailableNodes());
    return coreNodes;
}
```

</blockquote></details>

Done.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/util/PropertiesUtil.java, line 145 at r7](https://reviewable.io/reviews/openmpf/openmpf/517#-LEkBQAX-Uu3sOPuoyRm:-LEpWhZO5SzgXG2grPNu:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/2b4943534faac915400a80e8fb8f03d82b266749/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/util/PropertiesUtil.java#L145)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

As mentioned above I think we should make sure the collection doesn't get modified using either:
```
coreMpfNodes = ImmutableSet.copyOf(coreMpfNodesStr.split(","));
```
or
```
coreMpfNodes = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(coreMpfNodesStr.split(","))))
```
</blockquote></details>

Done.

---

*[trunk/workflow-manager/src/main/webapp/resources/mytheme/js/angular_app/directive_templates/page_info.html, line 9 at r7](https://reviewable.io/reviews/openmpf/openmpf/517#-LEkTx680uHWAKwfstSd:-LEpyKAR-CEzy1h6CnWA:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/2b4943534faac915400a80e8fb8f03d82b266749/trunk/workflow-manager/src/main/webapp/resources/mytheme/js/angular_app/directive_templates/page_info.html#L9)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

You can use the bootstrap `text-left` class here.
</blockquote></details>

Done.

---

*[trunk/workflow-manager/src/main/webapp/resources/mytheme/css/app.css, line 183 at r7](https://reviewable.io/reviews/openmpf/openmpf/517#-LEkTZx4CE-RNTJBOFyd:-LEpyL2S1sTgj5rozH33:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/2b4943534faac915400a80e8fb8f03d82b266749/trunk/workflow-manager/src/main/webapp/resources/mytheme/css/app.css#L183)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

Boostrap defines a `text-left` class that  you can use instead of this.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517)*
<!-- Sent from Reviewable.io -->

jrobble commented:
Review status: all files reviewed, 17 unresolved discussions (waiting on @brosenberg42)

---

*[trunk/bin/mpf-scripts/mpf/mpf_nodes.py, line 99 at r8](https://reviewable.io/reviews/openmpf/openmpf/517#-LEnq6cE2f56oONbpi1s:-LEpzCVuDFRWR4IRPmIU:b-d410vg) ([raw file](https://github.com/openmpf/openmpf/blob/c6ae7a56d0dd23c5e54494008e37204a33980eab/trunk/bin/mpf-scripts/mpf/mpf_nodes.py#L99)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I updated the code to return a list of strings without the 'u'. I don't quite like the idea of handling the value in a unique way was because it was generated using using `json.load`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517)*
<!-- Sent from Reviewable.io -->

brosenberg42 commented:
Reviewed 10 of 10 files at r10.
Review status: all files reviewed, 1 unresolved discussion (waiting on @jrobble)

---

*[trunk/workflow-manager/src/main/webapp/WEB-INF/html/admin/nodes/layout.html, line 49 at r10](https://reviewable.io/reviews/openmpf/openmpf/517#-LEttaTj8E0-eOWBZ1vB:-LEttaTj8E0-eOWBZ1vC:b-fsngju) ([raw file](https://github.com/openmpf/openmpf/blob/ac725b9a5a867eaff714e1a3ca3240d86a9770f8/trunk/workflow-manager/src/main/webapp/WEB-INF/html/admin/nodes/layout.html#L49)):*
> ```HTML
>                                 <span class="node-host" ng-if="!host.core">(Spare,</span>
>                                 <span class="node-host" ng-if="host.online">Online)</span>
>                                 <span class="node-host" ng-if="!host.online">Offline)</span>
> ```

You should disable the buttons for offline nodes. Currently you can still click on all the buttons for on offline node.

---


*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517)*
<!-- Sent from Reviewable.io -->

jrobble commented:
Still need to address issues with Sphinx tests failing on Jenkins.

---

Reviewed 2 of 2 files at r11.
Review status: all files reviewed, 1 unresolved discussion (waiting on @brosenberg42)

---

*[trunk/workflow-manager/src/main/webapp/WEB-INF/html/admin/nodes/layout.html, line 49 at r10](https://reviewable.io/reviews/openmpf/openmpf/517#-LEttaTj8E0-eOWBZ1vB:-LEuuJI338F1HPiCknD7:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/ac725b9a5a867eaff714e1a3ca3240d86a9770f8/trunk/workflow-manager/src/main/webapp/WEB-INF/html/admin/nodes/layout.html#L49)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

You should disable the buttons for offline nodes. Currently you can still click on all the buttons for on offline node.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/517#-:-LEuuLkT8k54qq-Gqphd:b-s1dkzm)*
<!-- Sent from Reviewable.io -->
